Sensing Subsystem Upstream PR1 (#55389)
zephyrproject-rtos/zephyr: doc/services/sensing/index.rst

Reviewer: OK, will fix it

* Decoupled with any host link protocols, it's Zephyr Application's role to handle different
  protocols (MQTT, HID or Private, all configurable)

Main Features
=============

* Scope
    * Focus on framework for sensor fusion, multiple clients, arbitration, data sampling, timing
      management and scheduling.

* Sensor Abstraction
    * ``Physical sensor``: interacts with Zephyr sensor device drivers, focus on data collecting.
    * ``Virtual sensor``: relies on other sensor(s), ``physical`` or ``virtual``, focus on
      data fusion.

* Data Driven Model
    * ``Polling mode``: periodical sampling rate
    * ``Interrupt mode``: data ready, threshold interrupt etc.

* Scheduling
    * single thread main loop for all sensor objects sampling and process.

* Buffer Mode for Batching

* Configurable Via Device Tree

API Design
**********

API Organization
================

* Sensing Subsystem
    * Sensor Types (:zephyr_file:`include/zephyr/sensing/sensing_sensor_types.h`)
    * Data Types (:zephyr_file:`include/zephyr/sensing/sensing_datatypes.h`)
    * Sensing Subsystem API (:zephyr_file:`include/zephyr/sensing/sensing.h`)

>     * Sensing Sensor API (:zephyr_file:`include/zephyr/sensing/sensing_sensor.h`)

Reviewer 2: Please don't just `resolve` comment.

.. _sensing_api:

Sensing Subsystem
########################

.. contents::
    :local:
    :depth: 2

Overview
********

Sensing Subsystem is a high level sensor framework inside the OS user
space service layer. It is a framework focused on sensor fusion, client
arbitration, sampling, timing, scheduling and sensor based power management.

Key concepts in Sensing Subsystem include physical sensor and virtual sensor objects,
and a scheduling framework over sensor object relationships.
Physical sensors do not depend on any other sensor objects for input, and
will directly interact with existing zephyr sensor device drivers.
Virtual sensors rely on other sensor objects (physical or virtual) as
report inputs.

The sensing subsystem relies on existing Zephyr sensor device APIs or V2

> zephyr sensor device APIs (`PR#44098 <https://github.com/zephyrproject-rtos/zephyr/pull/44098>`_),

Reviewer: Removed it, thanks for point out this!

Use of the sensing subsystem is optional. Applications that only need to access simple sensors
devices can use the Zephyr :ref:`sensor_api` API directly.

Since the sensing subsystem is separated from device driver layer or
kernel space and could support various customizations and sensor
algorithms in user space with virtual sensor concepts. The existing
sensor device driver can focus on low layer device side works, can keep
simple as much as possible, just provide device HW abstraction and
operations etc. This is very good for system stability.

The sensing subsystem is decoupled with any sensor expose/transfer
protocols, the target is to support various up-layer frameworks and
Applications with different sensor expose/transfer protocols,
such as `CHRE <https://github.com/zephyrproject-rtos/chre>`_, HID sensors Applications,
MQTT sensor Applications according different products requirements. Or even support multiple
Applications with different up-layer sensor protocols at the same time
with it's multiple clients support design. For example can support CHRE
and other normal zephyr sensor application (can use HID etc) at

> the same time.

Reviewer: Updated, it's more concise now, thanks a lot!

Sensing subsystem can help build a unified Zephyr sensing architecture for

> cross host OSes support and as well of IoT sensor solutions.

Reviewer: Fixed, thanks a lot!

The diagram below illustrates how the Sensing Subsystem integrates with up-layer frameworks.

.. image:: images/sensing_solution.png
   :align: center
   :alt: Unified Zephyr sensing architecture.

Configurability
===============

* Reusable and configurable standalone subsystem.
* Based on Zephyr existing low-level Sensor API (reuse 100+ existing sensor device drivers)
* Provide Zephyr high-level Sensing Subsystem API for Applications.
* Separate option CHRE Sensor PAL Implementation module to support CHRE.
* Decoupled with any host link protocols, it's Zephyr Application's role to handle different
  protocols (MQTT, HID or Private, all configurable)

Main Features
=============

* Scope
    * Focus on framework for sensor fusion, multiple clients, arbitration, data sampling, timing
      management and scheduling.

* Sensor Abstraction
    * ``Physical sensor``: interacts with Zephyr sensor device drivers, focus on data collecting.

>     * ``Virtual sensor``: relies on other sensor(s), ``physical`` or ``virtual``, focus on
>                           data fusion.

Reviewer 3: This bullet rendered with bold.  I think changing the indent of continuation line may fix it:
```suggestion
    * ``Virtual sensor``: relies on other sensor(s), ``physical`` or ``virtual``, focus on
      data fusion.
```

Reviewer: Fixed, thanks a lot!

* Data Driven Model
    * ``Polling mode``: periodical sampling rate
    * ``Interrupt mode``: data ready, threshold interrupt etc.

* Scheduling
    * single thread main loop for all sensor objects sampling and process.

* Buffer Mode for Batching

> * Configurable Via Device Tree

Reviewer 3: I think you need a blank line after each item for consistent rendering in HTML.

Reviewer: Updated, thanks a lot!

.. image:: images/sensing_api_org.png
   :align: center
   :alt: Sensing subsystem API organization.

``Sensing Subsystem API`` is for Applications.
``Sensing Sensor API`` is for development ``sensors``.

Major Flows
=========================

* Sensor Configuration Flow

.. image:: images/sensor_config_flow.png
   :align: center
   :alt: Sensor Configuration Flow (App set report interval to hinge angel sensor example).

* Sensor Data Flow

.. image:: images/sensor_data_flow.png
   :align: center
   :alt: Sensor Data Flow (App receive hinge angel data through data event callback example).

Sensor Types And Instance
=========================

The ``Sensing Subsystem`` supports multiple instances of the same sensor type,
there're two methods for Applications to identify and open an unique sensor instance:

* Enumerate all sensor instances

  :c:func:`sensing_get_sensors` returns all current board configuration supported sensor instances'
  information in a :c:struct:`sensing_sensor_info` pointer array .

  Then Applications can use :c:func:`sensing_open_sensor` to
  open specific sensor instance for future accessing, configuration and receive sensor data etc.

>   This method is suitable for supporting some up-layer frameworks like ``CHRE``, ``HID`` which need
>   dynamic enumerate underlying platforms' sensor instances.

Reviewer 3: ```suggestion
  This method is suitable for supporting some up-layer frameworks like ``CHRE``, ``HID`` which need
  to dynamically enumerate the underlying platforms' sensor instances.
```

Reviewer: Fixed, thanks a lot!

* Open the sensor instance by devicetree node directly

  Applications can use :c:func:`sensing_open_sensor_by_dt` to open a sensor instance directly with

>   sensor devicetree node name.

Reviewer 3: ```suggestion
  sensor devicetree node identifier.
```

Reviewer: Fixed, thanks a lot!

  For example:

.. code-block:: c

   sensing_open_sensor_by_dt(DEVICE_DT_GET(DT_NODELABLE(base_accel)), cb_list, handle);
   sensing_open_sensor_by_dt(DEVICE_DT_GET(DT_CHOSEN(zephyr_sensing_base_accel)), cb_list, handle);

> This method is useful and easy use for some simple Application which just want to access specific
>   sensor(s).

Reviewer 3: The extra indent causes a rendering error.

```suggestion
This method is useful and easy use for some simple Application which just want to access specific
sensor(s).
```

Reviewer: Fixed, thanks a lot!

> ``Sensor type`` follows the
>     `HID standard sensor types definition <https://usb.org/sites/default/files/hutrr39b_0.pdf>`_.

Reviewer 3: Same here, the extra spaces on the line continuation cause a rendering error.
```suggestion
``Sensor type`` follows the
`HID standard sensor types definition <https://usb.org/sites/default/files/hutrr39b_0.pdf>`_.
```

Reviewer: Fixed, thanks a lot!

Sensor Instance Handler
=========================

Clients using a ``void*`` type handler to handle a opened sensor
instance, and all subsequent operations on this sensor instance need use this handler,
such as set configurations, read sensor sample data, etc.

For a sensor instance, could have two kinds of clients:
``Application clients`` and ``Sensor clients``.

``Application clients`` can use :c:func:`sensing_open_sensor` to open a sensor instance
and get it's handler.

For ``Sensor clients``, there is no open API for opening a reporter, because the client-report
relationship is built at the sensor's registration stage with devicetree.

The ``Sensing Subsystem`` will auto open and create ``handlers`` for client sensor
to it's reporter sensors.
``Sensor clients`` can get it's reporters' handlers via :c:func:`sensing_sensor_get_reporters`.

.. image:: images/sensor_top.png
   :align: center
   :alt: Sensor Reporting Topology.

.. note::
   Sensors inside the Sensing Subsystem, the reporting relationship between them are all auto
   generated by Sensing Subsystem according devicetree definitions, handlers between client sensor
   and reporter sensors are auto created.

>    Application(s) need call sensing_sensor_open() APIs to explicitly open the sensor instance.

Reviewer 3: ```suggestion
   Application(s) need to call :c:func:`sensing_open_sensor to explicitly open the sensor instance.
```

Reviewer: Fixed! Thanks a lot!

Sensor Sample Value
==================================

* Data Structure

  Each sensor sample value defined as a common ``header`` + ``readings[]`` data structure:

  .. code-block:: c

      struct sensing_sensor_value_xxx {
         struct sensing_sensor_value_header header;
         struct data {
            uint32_t timestamp_delta;
            union {
                          ...
            };
         } readings[1];
      };

  The ``header`` definition:

  .. code-block:: c

      struct sensing_sensor_value_header {
         /** base timestamp of this data readings, unit is micro seconds */
         uint64_t base_timestamp;
         /** count of this data readings */
         uint16_t reading_count;

>       };

Reviewer 3: Use the `literalinclude::`  command to reference the actual `struct sensing_sensor_value_header` in the code.  That will ensure that the documentation stays up to date with the code.

Reviewer: Updated! Thanks a lot!

* Time Stamp

  Time stamp unit in sensing subsystem is ``micro seconds``.

>   The ``header`` defined a **base_timestamp**, and
>   each element in **readings[]** array defined **timestamp_delta**.

Reviewer 3: ```suggestion
  The ``header`` defines a **base_timestamp**, and
  each element in the **readings[]** array defines **timestamp_delta**.
```

Reviewer: Fixed, thanks a lot!

>   Here use **base_timestamp** (``uint64_t``) and **timestamp_delta** (``uint32_t``) to
>   save memory usage in batching mode.
> 
>   The **base_timestamp** is for ``readings[0]``, the **timestamp_delta** is relation
>   to the previous readings.

Reviewer 3: Suggestion to to use the text from the CHRE documentation.
```suggestion
  The **timestamp_delta** is in relation to the previous **readings** (or the **base_timestamp**)
  for ``readings[0]``. 
```

Reviewer: Updated, thanks a lot!

.. _sensing_api:

Sensing Subsystem
########################

.. contents::
    :local:
    :depth: 2

Overview
********

Sensing Subsystem is a high level sensor framework inside the OS user
space service layer. It is a framework focused on sensor fusion, client
arbitration, sampling, timing, scheduling and sensor based power management.

Key concepts in Sensing Subsystem include physical sensor and virtual sensor objects,
and a scheduling framework over sensor object relationships.
Physical sensors do not depend on any other sensor objects for input, and
will directly interact with existing zephyr sensor device drivers.
Virtual sensors rely on other sensor objects (physical or virtual) as
report inputs.

The sensing subsystem relies on existing zephyr sensor device APIs or V2

> zephyr sensor device APIs (`PR#44098 <https://github.com/zephyrproject-rtos/zephyr/pull/44098>`_).

Author: modified, thanks

So it can leverage current existing zephyr sensor device drivers (100+).
And it's configurable, so, for some low cost IoT devices, may not need an
advanced sensor framework, but just need access some simple sensor
devices,can not configure and build this sensing subsystem, but just use

> the exiting zephyr sensor device APIs to save memory resources.

Author: modified, thanks.

Since the sensing subsystem is separated from device driver layer or
kernel space and could support various customizations and sensor
algorithms in user space with virtual sensor concepts. The existing
sensor device driver can focus on low layer device side works, can keep
simple as much as possible, just provide device HW abstraction and
operations etc. This is very good for system stability.

The sensing subsystem is decoupled with any sensor expose/transfer
protocols, the target is to support various up-layer frameworks and
Applications with different sensor expose/transfer protocols,
such as `CHRE <https://github.com/zephyrproject-rtos/chre>`_, HID sensors Applications, MQTT sensor Applications
according different products requirements. Or even support multiple
Applications with different up-layer sensor protocols at the same time
with it's multiple clients support design. For example can support CHRE
and other normal zephyr sensor application (can use HID etc) at
the same time.

Sensing subsystem can help build a unified Zephyr sensing architecture for
cross host OSes support and as well of IoT sensor solutions.

> Below is the proposed solution with sensing subsystem:

Author: modified, thanks.

.. image:: images/sensing_solution.png
   :align: center
   :alt: Unified Zephyr sensing architecture.

Configurability
===============

* Reusable and configurable standalone subsystem.
* Based on Zephyr existing low-level Sensor API (reuse 100+ existing sensor device drivers)
* Provide Zephyr high-level Sensing Subsystem API for Applications.
* Separate option CHRE Sensor PAL Implementation module to support CHRE.
* Decoupled with any host link protocols, it's Zephyr Application's role to handle different
  protocols (MQTT, HID or Private, all configurable)

Configure Options
=================
* A: CHRE (optional, need config for Chrome host)
* B: CHRE Sensor PAL Implementation (optional, must if A configured)
* C: Sensing Subsystem (optional, must if B or D configured)

> * D: HID (optional, need config for Windows/Linux host)

Reviewer: OK, will remove it in this PR

> * D: HID (optional, need config for Windows/Linux host)

Author: removed it, thanks.

For simple low-cost sensor devices, can use existing Zephyr sensor driver API directly,  no

> need configure A, B, C or D.

Reviewer: OK, will delete it,  thanks!

> need configure A, B, C or D.

Author: modified, thanks.

``Sensor index`` 0 always indicate the default sensor instance of a ``sensor type``.

``Sensor type`` follows the `HID standard sensor types definition <https://usb.org/sites/default/files/hutrr39b_0.pdf>`_.

.. code-block:: c

    /**
     * sensor category light
     */
    #define SENSING_SENSOR_TYPE_LIGHT_AMBIENTLIGHT            0x41
    #define SENSING_SENSOR_TYPE_LIGHT_CONSUMER_INFRARED       0x42

    /**
     * sensor category motion
     */
    #define SENSING_SENSOR_TYPE_MOTION_ACCELEROMETER_1D       0x71
    #define SENSING_SENSOR_TYPE_MOTION_ACCELEROMETER_2D       0x72
    #define SENSING_SENSOR_TYPE_MOTION_ACCELEROMETER_3D       0x73
    #define SENSING_SENSOR_TYPE_MOTION_GYROMETER_1D           0x74
    #define SENSING_SENSOR_TYPE_MOTION_GYROMETER_2D           0x75
    #define SENSING_SENSOR_TYPE_MOTION_GYROMETER_3D           0x76
    #define SENSING_SENSOR_TYPE_MOTION_MOTION_DETECTOR        0x77
    #define SENSING_SENSOR_TYPE_MOTION_ACCELEROMETER          0x79
    #define SENSING_SENSOR_TYPE_MOTION_GYROMETER              0x7A

    /**
     * sensor category other
     */
    #define SENSING_SENSOR_TYPE_OTHER_CUSTOM                  0xE1

>     #define SENSING_SENSOR_TYPE_ALL                           0xFFFF

Reviewer: Yes, already used `zephyr_file` to point to `sensing_sensor_types.h`
